### PR TITLE
Adjusts velocity boundary conditions for accelerated flows

### PR DESCRIPTION
**Pull request by marinlauber**
I have adjusted the velocity boundary conditions to allow for either a `Tuple` or a `Function(i,t)`. This is required if we have an accelerated flow without periodic BC in the forced direction. The resulting body force is computed with `ForwardDiff` on the body velocity function.

Either boundary condition type is dispatched to a `BCTuple` function that generates a Tuple that is then passed to the `BC!()` function. 

I added an example, `examples/TwoD_SlowStartCircle.jl`, that demonstrates how to use the time-varying velocity boundary conditions.

I have added some tests (not run on the GPU yet) for the new functions.

### CODE REVIEW

**Comment by b-fg** [comment]
> This requires storing the velocity BC as an Array instead of a Tuple. I am not sure if this needs to be stored on the GPU or not (I will test this).

That's indeed required for the GPU. We cannot have a small `CuArray` and access it by index, so this is why tuples are used instead.

File: src/Flow.jl [hunk 1]
```diff
@@ -59,17 +59,21 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
...
 If `g ≠ nothing`, then `g(i,t)=dUᵢ/dt`.
 """
-accelerate!(r,t,g) = for i ∈ 1:last(size(r))
+accelerate!(U,r,t,g,::Nothing) = for i ∈ 1:last(size(r))
     r[..,i] .+= g(i,t)
 end
-accelerate!(r,t,::Nothing) = nothing
+accelerate!(U,r,t,::Nothing,Uₜ::Function) = for i ∈ 1:last(size(r))
+    r[..,i] .+= ForwardDiff.derivative(τ->Uₜ(i,τ),t)
+    U[i] = Uₜ(i,t)
```
Comment (marinlauber) [inline src/Flow.jl:74]: This line is causing the trouble; it works fine on the CPU, but the GPU complains.

**Comment by b-fg** [comment]
Have a look here: https://juliaarrays.github.io/StaticArrays.jl/stable/pages/api/#Mutable-arrays:-MVector,-MMatrix-and-MArray

I think `MVector` structures is what you are looking for.

**Comment by weymouth** [comment]
Shouldn't we just use functions instead of arrays? @TzuYaoHuang's `g` function works fine on GPUs. Why do we need an array or Tuple at all?

**Comment by b-fg** [comment]
Maybe we can dispatch to the tuple for static conditions and dispatch to the functions for dynamic ones.

File: src/Flow.jl [hunk 2]
```diff
@@ -66,10 +66,16 @@ using EllipsisNotation
...
-accelerate!(r,t,g) = for i ∈ 1:last(size(r))
+accelerate!(r,t,g::Function,::Tuple) = for i ∈ 1:last(size(r))
     r[..,i] .+= g(i,t)
 end
-accelerate!(r,t,::Nothing) = nothing
+accelerate!(r,t,::Nothing,U::Function) = for i ∈ 1:last(size(r))
+    r[..,i] .+= ForwardDiff.derivative(τ->U(i,τ),t)
+end
+accelerate(r,t,::Function,::Function) = for i ∈ 1:last(size(r))
+    r[..,i] .+= g(i,t) + ForwardDiff.derivative(τ->U(i,τ),t)
```
Comment (b-fg) [inline src/Flow.jl:76]: `U` seems to not be in the scope. Also this is not covered by tests which is why it is not being picked up as a test error.

File: src/Flow.jl [hunk 3]
```diff
@@ -93,19 +99,19 @@ struct Flow{D, T, Sf<:AbstractArray{T}, Vf<:AbstractArray{T}, Tf<:AbstractArray{
     μ₀:: Vf # zeroth-moment vector
     μ₁:: Tf # first-moment tensor field
     # Non-fields
-    U :: NTuple{D, T} # domain boundary values
+    U :: Union{NTuple{D,Number},Function} # domain boundary values
     Δt:: Vector{T} # time step (stored in CPU memory)
     ν :: T # kinematic viscosity
     g :: Union{Function,Nothing} # (possibly time-varying) uniform acceleration field
     exitBC :: Bool # Convection exit
     perdir :: NTuple # direction of periodic direction
```
Comment (b-fg) [inline src/Flow.jl:107]:
```suggestion
    perdir :: NTuple # tuple of periodic directions
```

Comment (b-fg) [reply src/Flow.jl:107]: (I know this is from another PR, but I just spotted it)

File: src/Flow.jl (hunk 2 above)
Comment (marinlauber) [reply src/Flow.jl:76]: Neither is `g` actually. Luckily I didn't test it ;)

Comment (marinlauber) [reply src/Flow.jl:76]: And the name is `accelerate()`, not `accelerate!()`...

File: test/runtests.jl [hunk 4]
```diff
@@ -163,12 +166,16 @@ end
 
     # check applying acceleration
     N = 4
-    a = zeros(N,N,2)
-    WaterLily.accelerate!(a,1,nothing)
+    a = zeros(N,N,2);
```
Comment (b-fg) [inline test/runtests.jl:169]: `;` why

**Comment by b-fg** [comment]
I think this is ready to merge? @weymouth @marinlauber

**Comment by marinlauber** [comment]
This is ready!

The tests are passing on CPU/GPU

```bash
Test Summary: | Pass  Total   Time
util.jl       |   34     34  20.8s
Test Summary: | Pass  Total   Time
Poisson.jl    |   14     14  18.8s
Test Summary:        | Pass  Total  Time
MultiLevelPoisson.jl |   13     13  6.2s
Test Summary: | Pass  Total  Time
Flow.jl       |   18     18  9.4s
Test Summary: | Pass  Total  Time
Body.jl       |    3      3  0.1s
Test Summary: | Pass  Total  Time
AutoBody.jl   |   11     11  5.3s
Test Summary:        | Pass  Total  Time
Flow.jl periodic TGV |    2      2  5.1s
Test Summary:                      | Pass  Total   Time
Flow.jl with increasing body force |    2      2  10.0s
Test Summary: | Pass  Total  Time
Metrics.jl    |   17     17  6.8s
Test Summary: | Pass  Total  Time
WaterLily.jl  |   34     34  6.4s
Test Summary: | Pass  Total   Time
VTKExt.jl     |   28     28  14.2s
     Testing WaterLily tests passed
``` 

And the benchmark results compared to `master (d0ec1d2)`.

```bash
--------------------------------------
Running benchmark tests for:
 - Julia:        1.10.0
 - Backends:     Array CuArray
 - CPU threads:  1 6
 - Cases:        tgv jelly
 - Size:         6,7 5,6
 - Sim. steps:   100 100
 - Data type:    Float32 Float32
--------------------------------------

Julia 1.10.0 benchmaks
Updating environment to Julia 1.10.0
   Resolving package versions...
  No Changes to `~/Workspace/WaterLily.jl/benchmark/Project.toml`
  No Changes to `~/Workspace/WaterLily.jl/benchmark/Manifest.toml`
    Updating registry at `/scratch/marinlauber/.julia/registries/General.toml`
  No Changes to `~/Workspace/WaterLily.jl/benchmark/Project.toml`
  No Changes to `~/Workspace/WaterLily.jl/benchmark/Manifest.toml`
Running: julia --project --startup-file=no -t 1 benchmark.jl --backend=Array --cases=["tgv","jelly"] --log2p=[(6,7),(5,6)] --max_steps=[100,100] --ftype=[Float32,Float32]
Benchmarking: tgv
(1/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 9.153122932 seconds)
done (took 9.664080328 seconds)
(2/2) benchmarking "7"...
  (1/1) benchmarking "sim_step!"...
  done (took 55.775909505 seconds)
done (took 56.219365421 seconds)
Benchmarking: jelly
(1/2) benchmarking "5"...
  (1/1) benchmarking "sim_step!"...
  done (took 10.31607692 seconds)
done (took 10.762646285 seconds)
(2/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 48.866873048 seconds)
done (took 49.323129197 seconds)
Running: julia --project --startup-file=no -t 6 benchmark.jl --backend=Array --cases=["tgv","jelly"] --log2p=[(6,7),(5,6)] --max_steps=[100,100] --ftype=[Float32,Float32]
Benchmarking: tgv
(1/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 3.224376419 seconds)
done (took 3.562162138 seconds)
(2/2) benchmarking "7"...
  (1/1) benchmarking "sim_step!"...
  done (took 15.920800915 seconds)
done (took 16.161406035 seconds)
Benchmarking: jelly
(1/2) benchmarking "5"...
  (1/1) benchmarking "sim_step!"...
  done (took 3.995909223 seconds)
done (took 4.262310484 seconds)
(2/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 18.140882184 seconds)
done (took 18.3939807 seconds)
Running: julia --project --startup-file=no benchmark.jl --backend=CuArray --cases=["tgv","jelly"] --log2p=[(6,7),(5,6)] --max_steps=[100,100] --ftype=[Float32,Float32]
Benchmarking: tgv
(1/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 1.097216626 seconds)
done (took 1.657903387 seconds)
(2/2) benchmarking "7"...
  (1/1) benchmarking "sim_step!"...
  done (took 1.639622323 seconds)
done (took 2.123524882 seconds)
Benchmarking: jelly
(1/2) benchmarking "5"...
  (1/1) benchmarking "sim_step!"...
  done (took 1.510458438 seconds)
done (took 2.007634304 seconds)
(2/2) benchmarking "6"...
  (1/1) benchmarking "sim_step!"...
  done (took 2.660584546 seconds)
done (took 3.166380567 seconds)
```

```bash
Benchmark environment: tgv sim_step! (max_steps=100)
▶ log2p = 6
┌─────────┬───────────┬────────┬───────────┬─────────────┬────────┬──────────┬──────────┐
│ Backend │ WaterLily │ Julia  │ Precision │ Allocations │ GC [%] │ Time [s] │ Speed-up │
├─────────┼───────────┼────────┼───────────┼─────────────┼────────┼──────────┼──────────┤
│   CPUx1 │   afa03a3 │ 1.10.0 │   Float32 │     1271464 │   0.09 │     8.67 │     1.00 │
│   CPUx1 │   d0ec1d2 │ 1.10.0 │   Float32 │     1357166 │   0.23 │     8.95 │     0.97 │
│   CPUx6 │   afa03a3 │ 1.10.0 │   Float32 │     4677363 │   1.00 │     2.92 │     2.97 │
│   CPUx6 │   d0ec1d2 │ 1.10.0 │   Float32 │     4999132 │   0.94 │     3.13 │     2.77 │
│     GPU │   afa03a3 │ 1.10.0 │   Float32 │     2304525 │   1.56 │     0.57 │    15.17 │
│     GPU │   d0ec1d2 │ 1.10.0 │   Float32 │     2467970 │   1.58 │     0.60 │    14.57 │
└─────────┴───────────┴────────┴───────────┴─────────────┴────────┴──────────┴──────────┘
▶ log2p = 7
┌─────────┬───────────┬────────┬───────────┬─────────────┬────────┬──────────┬──────────┐
│ Backend │ WaterLily │ Julia  │ Precision │ Allocations │ GC [%] │ Time [s] │ Speed-up │
├─────────┼───────────┼────────┼───────────┼─────────────┼────────┼──────────┼──────────┤
│   CPUx1 │   afa03a3 │ 1.10.0 │   Float32 │     1145360 │   0.09 │    55.34 │     1.00 │
│   CPUx1 │   d0ec1d2 │ 1.10.0 │   Float32 │     1329041 │   0.14 │    57.37 │     0.96 │
│   CPUx6 │   afa03a3 │ 1.10.0 │   Float32 │     3994709 │   0.40 │    15.65 │     3.54 │
│   CPUx6 │   d0ec1d2 │ 1.10.0 │   Float32 │     4683048 │   0.67 │    16.54 │     3.35 │
│     GPU │   afa03a3 │ 1.10.0 │   Float32 │     1918194 │   0.38 │     1.16 │    47.86 │
│     GPU │   d0ec1d2 │ 1.10.0 │   Float32 │     2263519 │   0.68 │     1.31 │    42.10 │
└─────────┴───────────┴────────┴───────────┴─────────────┴────────┴──────────┴──────────┘
```

```bash
Benchmark environment: jelly sim_step! (max_steps=100)
▶ log2p = 5
┌─────────┬───────────┬────────┬───────────┬─────────────┬────────┬──────────┬──────────┐
│ Backend │ WaterLily │ Julia  │ Precision │ Allocations │ GC [%] │ Time [s] │ Speed-up │
├─────────┼───────────┼────────┼───────────┼─────────────┼────────┼──────────┼──────────┤
│   CPUx1 │   afa03a3 │ 1.10.0 │   Float32 │     2263424 │   0.15 │     9.86 │     1.00 │
│   CPUx1 │   d0ec1d2 │ 1.10.0 │   Float32 │     2701319 │   0.17 │    10.94 │     0.90 │
│   CPUx6 │   afa03a3 │ 1.10.0 │   Float32 │     8318725 │   1.44 │     3.72 │     2.65 │
│   CPUx6 │   d0ec1d2 │ 1.10.0 │   Float32 │     9932893 │   1.35 │     4.29 │     2.30 │
│     GPU │   afa03a3 │ 1.10.0 │   Float32 │     4202842 │   1.23 │     1.00 │     9.87 │
│     GPU │   d0ec1d2 │ 1.10.0 │   Float32 │     5081193 │   1.40 │     1.18 │     8.34 │
└─────────┴───────────┴────────┴───────────┴─────────────┴────────┴──────────┴──────────┘
▶ log2p = 6
┌─────────┬───────────┬────────┬───────────┬─────────────┬────────┬──────────┬──────────┐
│ Backend │ WaterLily │ Julia  │ Precision │ Allocations │ GC [%] │ Time [s] │ Speed-up │
├─────────┼───────────┼────────┼───────────┼─────────────┼────────┼──────────┼──────────┤
│   CPUx1 │   afa03a3 │ 1.10.0 │   Float32 │     3192831 │   0.19 │    48.40 │     1.00 │
│   CPUx1 │   d0ec1d2 │ 1.10.0 │   Float32 │     4208670 │   0.22 │    53.38 │     0.91 │
│   CPUx6 │   afa03a3 │ 1.10.0 │   Float32 │    11729531 │   0.84 │    17.86 │     2.71 │
│   CPUx6 │   d0ec1d2 │ 1.10.0 │   Float32 │    15448298 │   0.88 │    20.85 │     2.32 │
│     GPU │   afa03a3 │ 1.10.0 │   Float32 │     5958203 │   0.89 │     2.15 │    22.48 │
│     GPU │   d0ec1d2 │ 1.10.0 │   Float32 │     7945566 │   1.02 │     2.69 │    17.99 │
└─────────┴───────────┴────────┴───────────┴─────────────┴────────┴──────────┴──────────┘
```

File: src/Flow.jl [hunk 5]
```diff
@@ -137,19 +143,21 @@ end
...
+@fastmath function mom_step!(a::Flow{N},b::AbstractPoisson) where N
     a.u⁰ .= a.u; scale_u!(a,0)
     # predictor u → u'
+    U = BCTuple(a.U,time(a),N)
     conv_diff!(a.f,a.u⁰,a.σ,ν=a.ν,perdir=a.perdir)
-    accelerate!(a.f,time(a),a.g)
-    BDIM!(a); BC!(a.u,a.U,a.exitBC,a.perdir)
-    a.exitBC && exitBC!(a.u,a.u⁰,a.U,a.Δt[end]) # convective exit
-    project!(a,b); BC!(a.u,a.U,a.exitBC,a.perdir)
+    accelerate!(a.f,time(a),a.g,a.U)
```
Comment (b-fg) [inline src/Flow.jl:151]: is the last argument supposed to be `a.U` or `U`?

File: src/Flow.jl [hunk 6]
```diff
@@ -137,19 +143,21 @@ end
...
+    BDIM!(a); BC!(a.u,U,a.exitBC,a.perdir)
+    a.exitBC && exitBC!(a.u,a.u⁰,U,a.Δt[end]) # convective exit
+    project!(a,b); BC!(a.u,U,a.exitBC,a.perdir)
     # corrector u → u¹
+    U = BCTuple(a.U,timeNext(a),N)
     conv_diff!(a.f,a.u,a.σ,ν=a.ν,perdir=a.perdir)
-    accelerate!(a.f,timeNext(a),a.g)
-    BDIM!(a); scale_u!(a,0.5); BC!(a.u,a.U,a.exitBC,a.perdir)
-    project!(a,b,0.5); BC!(a.u,a.U,a.exitBC,a.perdir)
+    accelerate!(a.f,timeNext(a),a.g,a.U)
```
Comment (b-fg) [inline src/Flow.jl:158]: and same here

File: test/runtests.jl [hunk 7]
```diff
@@ -164,10 +167,14 @@ end
     # check applying acceleration
     N = 4
     a = zeros(N,N,2)
```
Comment (b-fg) [inline test/runtests.jl:169]: this test is hardcoding `Array` only. Have you checked tests with `a::CuArray`?

File: src/Flow.jl (hunk 5 above)
Comment (marinlauber) [reply src/Flow.jl:151]: Yes, this is correct. It sets the correct type dispatch for `accelerate`

File: src/Flow.jl (hunk 6 above)
Comment (marinlauber) [reply src/Flow.jl:158]: Same as above.

File: test/runtests.jl [hunk 8]
```diff
@@ -165,17 +165,17 @@ end
...
-    @test all(a[:,:,1] .== 1) && all(a[:,:,2] .== 2)
-    WaterLily.accelerate!(a,1,nothing,(i,t) -> i==1 ? -t : -2*t)
-    @test all(a[:,:,1] .== 0) && all(a[:,:,2] .== 0)
-    WaterLily.accelerate!(a,1,(i,t) -> i==1 ? t : 2*t,(i,t) -> i==1 ? -t : -2*t)
-    @test all(a[:,:,1] .== 0) && all(a[:,:,2] .== 0)
-
+    for f ∈ arrays
+        N = 4; a = zeros(N,N,2) |> f
+        WaterLily.accelerate!(a,1,nothing,())
+        @test all(a .== 0)
```
Comment (b-fg) [inline test/runtests.jl:171]: missing the `allowscalar` for all these tests?

File: examples/TwoD_SlowStartCircle.jl [hunk 9]
```diff
@@ -0,0 +1,12 @@
+using WaterLily
+using CUDA
+function circle(m,n;a0=0.5,Re=250,U=1,f=Array)
+    # define a circle at the domain center
+    radius = n/8
+    body = AutoBody((x,t)->√sum(abs2, x .- (m/4,n/2)) - radius)
+    # define time-varying velocity boundary conditions
+    Ut(i,t::T) where {T} = i==1 ? convert(T,a0*t+(1.0+tanh(31.4*(t-1.0/a0)))/2.0*(1-a0*t)) : convert(T,0.0)
```
Comment (b-fg) [inline examples/TwoD_SlowStartCircle.jl:8]:
```suggestion
    Ut(i,t::T) where T = i==1 ? convert(T,a0*t+(1.0+tanh(31.4*(t-1.0/a0)))/2.0*(1-a0*t)) : zero(T)
```

File: test/runtests.jl (hunk 8 above)
Comment (marinlauber) [reply test/runtests.jl:171]: I double-checked, and the GPU only complains if I do `a[1,1,1]==0`, he is happy in this case :)

**Comment by marinlauber** [comment]
I think this is ready to merge, or do you want me to change other things? @b-fg @weymouth

**Comment by b-fg** [comment]
Looks fine to me.
